base/inference.jl

```diff
@@ -4731,10 +4730,17 @@ function statement_cost(ex::Expr, src::CodeInfo, mod::Module, params::InferenceP
         return ex.typ == Union{} ? 0 : plus_saturate(20, argcost)
     elseif head == :llvmcall
         return plus_saturate(10, argcost) # a wild guess at typical cost
-    elseif (head == :&)
-        return plus_saturate(length(ex.args), argcost)
+    elseif head == :enter
+        # try/catch is a couple function calls,
+        # but don't inline functions with try/catch
+        # since these aren't usually performance-sensitive functions,
+        # and llvm is more likely to miscompile them when these functions get large
+        return typemax(Int)
+    elseif head == :gotoifnot
+        target = ex.args[2]
+        return plus_saturate(target < line ? 40 : 5, argcost) # loops are more expensive than simple branches
```

Comment:
Interesting, this is a less-extreme variant of an early version I had where I disabled inlining on any function with a loop (and which exhibited performance problems because it was too conservative). This seems much better. With a cost of 40, it might be worth considering bumping the inlining threshold of 100 up a little higher? 100 is roughly the time of 5 calls, and often if we could make a 20% performance improvement we'd take it. I think I had it set to 200 initially, but lowered it to 100 so as to essentially match the system-image size. If these changes solve the image-size increase, then I suspect it would be better to go back to 200.